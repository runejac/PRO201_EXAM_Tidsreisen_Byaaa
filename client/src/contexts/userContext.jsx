import React, { createContext } from "react";
import { fetchJSON_client, putJSON } from "../helpers/http.jsx";

export const UserContext = createContext({
  async getUser() {
    return await fetchJSON_client("/api/login");
  },
  async updateUser(user) {
    return await putJSON("/api/login/updateuser", user);
  },
});
