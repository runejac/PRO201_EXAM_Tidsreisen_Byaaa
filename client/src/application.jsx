import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginPage } from "./pages/LoginPage.jsx";
import { MapPage } from "./pages/MapPage.jsx";
import { Quiz } from "./pages/capsules/quizCapsule/Quiz.jsx";
import { History } from "./pages/capsules/HistoryCapsule/History.jsx";
import { MyFindings } from "./pages/MyFindings.jsx";
import { UserContext } from "./contexts/userContext.jsx";
import { Profile } from "./pages/Profile.jsx";
import { NotFound } from "./components/NotFound.jsx";
import { BackButton } from "./components/BackButton";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useLoading } from "./helpers/useLoading.jsx";
import WithoutNavbar from "./components/WithoutNavbar";
import WithNavbar from "./components/WithNavbar";
import Audio from "./pages/capsules/AudioCapsule/Audio.jsx";
import Camera from "./components/Camera";
import WithBackButton from "./components/WithBackButton.jsx";

export const User = createContext("");
export const MapContext = React.createContext(null);

export function Application() {
  const [user, setUser] = useState([]);
  const [map, setMap] = useState(null);

  const { getUser } = useContext(UserContext);
  const { data: username, reload, loading, error } = useLoading(getUser);

  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const mapProvider = useMemo(() => ({ map, setMap }), [map, setMap]);

  useEffect(() => {
    username ? setUser(...username, username) : reload();
  }, [username]);

  return (
    <User.Provider value={providerValue}>
      <MapContext.Provider value={mapProvider}>
        <BrowserRouter>
          <main>
            <Routes>
              <Route element={<WithoutNavbar />}>
                <Route path="/" element={<LoginPage />} />
              </Route>
              <Route element={<WithNavbar />}>
                <Route path="/map" element={<MapPage />} />
                <Route path="/camera" element={<Camera />} />
                <Route element={<WithBackButton />}>
                  <Route path="/quiz/:id" element={<Quiz />} />
                  <Route path="/history/:id" element={<History />} />
                  <Route
                    path="/myfindings"
                    element={<MyFindings loading={loading} error={error} />}
                  />
                  <Route path="/audio/:id" element={<Audio />} />
                  <Route path="/profile" element={<Profile />} />
                  <Route path={"*"} element={<NotFound />} />
                </Route>
              </Route>
            </Routes>
          </main>
        </BrowserRouter>
      </MapContext.Provider>
    </User.Provider>
  );
}
