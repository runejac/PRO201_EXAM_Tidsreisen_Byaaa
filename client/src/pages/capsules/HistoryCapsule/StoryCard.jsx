import React, { useContext, useEffect, useState } from "react";
import { Loading } from "../../../components/Loading.jsx";
// swiper bundle styles
import "swiper/swiper-bundle.min.css";

// swiper core styles
import "swiper/swiper.min.css";

// modules styles
import "swiper/components/navigation/navigation.min.css";
import "swiper/components/pagination/pagination.min.css";

import { Swiper, SwiperSlide } from "swiper/react";
import { Scrollbar } from "swiper";
import SwiperCore from "swiper";
import { useParams } from "react-router-dom";
import "./storycard.css";
import { UserContext } from "../../../contexts/userContext.jsx";
import { HistoryDone } from "./HistoryDone.jsx";
import { User } from "../../../application.jsx";
import { ErrorModal } from "../../../components/ErrorModal.jsx";
import { FaBook } from "react-icons/fa";

export function StoryCard({ historyCapsule, error, loading }) {
  SwiperCore.use([Scrollbar]);

  const { id } = useParams();
  const { updateUser } = useContext(UserContext);
  const { user, setUser } = useContext(User);
  const { name, intro, walk, points } = user;
  const [count, setCount] = useState(0);
  let previousState = { ...user };

  // setting capsule object to be inserted in database
  const capsuleObject = {
    id: historyCapsule.id,
    name: historyCapsule.name,
    category: historyCapsule.category,
  };

  // updates user with capsuleObject and points
  const updateToDatabase = async () => {
    setCount(count + 1);
    await updateUser({
      user,
      finishedCapsules: capsuleObject,
      points: 20,
    });
  };

  // updates correct user only, and only if count is more than 0
  // counts when swiping
  useEffect(() => {
    if (count > 0) {
      return () => {
        if (!user.finishedCapsules.includes(capsuleObject)) {
          user.finishedCapsules.push(capsuleObject);
          user.points = user.points + 20;
          setUser({
            ...previousState,
          });
        }
      };
    }
  }, [count]);

  // showing correct history capsule according to what the url parameter is
  if (historyCapsule.name.toLowerCase() === id?.toLowerCase()) {
    return (
      <div>
        <div className={"p-4"} style={{ position: "relative", zIndex: "0" }}>
          <div style={{ position: "relative" }}>
            <h3 id={"capsule-name"} className={"text-center"}>
              <FaBook color={"var(--textColorGray)"} /> Historiekapsel
            </h3>
            <h1
              className={"text-center"}
              style={{
                color: "var(--backgroundColorGreeny)",
                fontFamily: "Source Sans Pro Bold",
              }}
            >
              {historyCapsule.name}
            </h1>
          </div>
          <Swiper
            centeredSlides={true}
            spaceBetween={50}
            slidesPerView={1}
            scrollbar={{ draggable: true }}
            onReachEnd={async () => await updateToDatabase()}
          >
            {historyCapsule.story.map((capsuleStory, index) => {
              return (
                <SwiperSlide key={index} className={"slide"}>
                  {capsuleStory.done === true ? (
                    <HistoryDone
                      updateToDatabase={updateToDatabase}
                      id={historyCapsule.id}
                    />
                  ) : (
                    <div className={"swiper-slide"} key={index}>
                      <div className={"slide-content"}>
                        <div id={"slide-container"}>
                          <div
                            /*style={{ maxWidth: "45rem" }}*/
                            className={"mt-5 image-container"}
                          >
                            {capsuleStory.image && (
                              <section id={"image-section"}>
                                <img
                                  id={"history-image"}
                                  className={"card-img"}
                                  src={capsuleStory.image}
                                  alt={"bilde av historie-element"}
                                />
                              </section>
                            )}
                          </div>

                          <section id={"story-section"} className={"mt-4"}>
                            <p id={"story-paragraph"}>{capsuleStory.story}</p>
                          </section>
                        </div>
                      </div>
                      <div id={"year"}>
                        <h4 className=" fw-bold">{capsuleStory.year}</h4>
                      </div>
                    </div>
                  )}
                </SwiperSlide>
              );
            })}
          </Swiper>
        </div>
      </div>
    );
  }
  return (
    <>
      {loading && <Loading />}
      {error && <ErrorModal error={error} />}
    </>
  );
}
