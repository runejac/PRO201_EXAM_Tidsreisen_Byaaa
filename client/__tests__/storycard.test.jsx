/**
 * @jest-environment jsdom
 */

import React from "react";
import { MemoryRouter, ReactRouter } from "react-router-dom";
import { fireEvent, render } from "@testing-library/react";
import { User } from "../src/application.jsx";
import { UserContext } from "../src/contexts/userContext.jsx";
import { StoryCard } from "../src/pages/capsules/HistoryCapsule/StoryCard.jsx";

//TODO: Se linje 43-44, 53-57, 106, og 152 i StoryCard.jsx

jest.mock("react-router-dom", () => ({
  useParams: () => ({
    id: "Test",
  }),
}));

describe("StoryCard", () => {
  it("shows snapshot", async () => {
    const user = {
      name: "Hei",
      points: 23,
      finishedCapsules: [],
      intro: true,
      walk: false,
    };
    const capsule = {
      id: 2,
      name: "Test",
      category: "TestKategori",
      story: [
        {
          year: "Ã…r 2022",
          story: "Dette er en test",
          image: "",
          done: false,
        },
      ],
    };
    const context = jest.fn();
    const component = render(
      <User.Provider value={{ user }}>
        <UserContext.Provider value={{ context }}>
          <StoryCard historyCapsule={capsule} />
        </UserContext.Provider>
      </User.Provider>
    );
    expect(component).toMatchSnapshot();
  });
  jest.clearAllMocks();
});
