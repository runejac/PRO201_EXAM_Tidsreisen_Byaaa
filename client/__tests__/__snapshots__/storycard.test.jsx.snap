// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryCard shows snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="p-4"
          style="position: relative; z-index: 0;"
        >
          <div
            style="position: relative;"
          >
            <h3
              class="text-center"
              id="capsule-name"
            >
              <svg
                color="var(--textColorGray)"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M448 360V24c0-13.3-10.7-24-24-24H96C43 0 0 43 0 96v320c0 53 43 96 96 96h328c13.3 0 24-10.7 24-24v-16c0-7.5-3.5-14.3-8.9-18.7-4.2-15.4-4.2-59.3 0-74.7 5.4-4.3 8.9-11.1 8.9-18.6zM128 134c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm0 64c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm253.4 250H96c-17.7 0-32-14.3-32-32 0-17.6 14.4-32 32-32h285.4c-1.9 17.1-1.9 46.9 0 64z"
                />
              </svg>
               Historiekapsel
            </h3>
            <h1
              class="text-center"
              style="font-family: Source Sans Pro Bold;"
            >
              Test
            </h1>
          </div>
          <div
            class="swiper-container swiper-container-initialized swiper-container-horizontal"
          >
            <div
              class="swiper-scrollbar"
            >
              <div
                class="swiper-scrollbar-drag"
                style="transform: translate3d(NaNpx, 0, 0);"
              />
            </div>
            <div
              class="swiper-wrapper"
              style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
            >
              <div
                class="swiper-slide slide"
              >
                <div
                  class="swiper-slide"
                >
                  <div
                    class="slide-content"
                  >
                    <div
                      id="slide-container"
                    >
                      <div
                        class="mt-5 image-container"
                      />
                      <section
                        class="mt-4"
                        id="story-section"
                      >
                        <p
                          id="story-paragraph"
                        >
                          Dette er en test
                        </p>
                      </section>
                    </div>
                  </div>
                  <div
                    id="year"
                  >
                    <h4
                      class=" fw-bold"
                    >
                      År 2022
                    </h4>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="p-4"
        style="position: relative; z-index: 0;"
      >
        <div
          style="position: relative;"
        >
          <h3
            class="text-center"
            id="capsule-name"
          >
            <svg
              color="var(--textColorGray)"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M448 360V24c0-13.3-10.7-24-24-24H96C43 0 0 43 0 96v320c0 53 43 96 96 96h328c13.3 0 24-10.7 24-24v-16c0-7.5-3.5-14.3-8.9-18.7-4.2-15.4-4.2-59.3 0-74.7 5.4-4.3 8.9-11.1 8.9-18.6zM128 134c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm0 64c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm253.4 250H96c-17.7 0-32-14.3-32-32 0-17.6 14.4-32 32-32h285.4c-1.9 17.1-1.9 46.9 0 64z"
              />
            </svg>
             Historiekapsel
          </h3>
          <h1
            class="text-center"
            style="font-family: Source Sans Pro Bold;"
          >
            Test
          </h1>
        </div>
        <div
          class="swiper-container swiper-container-initialized swiper-container-horizontal"
        >
          <div
            class="swiper-scrollbar"
          >
            <div
              class="swiper-scrollbar-drag"
              style="transform: translate3d(NaNpx, 0, 0);"
            />
          </div>
          <div
            class="swiper-wrapper"
            style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
          >
            <div
              class="swiper-slide slide"
            >
              <div
                class="swiper-slide"
              >
                <div
                  class="slide-content"
                >
                  <div
                    id="slide-container"
                  >
                    <div
                      class="mt-5 image-container"
                    />
                    <section
                      class="mt-4"
                      id="story-section"
                    >
                      <p
                        id="story-paragraph"
                      >
                        Dette er en test
                      </p>
                    </section>
                  </div>
                </div>
                <div
                  id="year"
                >
                  <h4
                    class=" fw-bold"
                  >
                    År 2022
                  </h4>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
